source/common/tls/connection_info_impl_base.cc

```diff
@@ -17,6 +17,16 @@ namespace Extensions {
 namespace TransportSockets {
 namespace Tls {
 
+namespace {
+// There must be an version of this function for each type possible in variant `CachedValue`.
+bool shouldRecalculateCachedEntry(const std::string& str) { return str.empty(); }
+bool shouldRecalculateCachedEntry(const std::vector<std::string>& vec) { return vec.empty(); }
```

Comment:
For URI SANs, this would recalculate the values every time if certificate has no URI SANs. Can we use optionals to store the value?

Reply:
I thought about that, and decided that the cost to recalculate the empty list is extremely low, and I opted for code simplicity.

I thought about using optionals, but the outer functions all return a reference, so we need a concrete empty object, not an empty optional, to return, so I would have to store an extra boolean for "this one is empty but should be kept". Or possibly keep a `static const` empty value for each type, and return a reference to that. 

My opinion is that this is good enough, and this is also parity with the behavior before I optimized the memory usage; I looked at the history, and it was always re-evaluating if the cached URI SANs vector was empty.

Reply:
OK, it's fine in the absence of a direct hook for the handshake completion.